neps/nep-0539.md

```diff
@@ -0,0 +1,741 @@
+for a transaction from when it is accepted to when it finishes execution becomes
+larger and larger, as the receipts need to queue behind those already in the
+system.
+
+First attempts to limit the memory consumption have been added without protocol
+changes. This is known as "local congestion control" and can be summarized in
+two rules:
+
+- Limit the transaction pool to 100 MB.
+  https://github.com/near/nearcore/pull/9172
+- Once we have accumulated more than 20k delayed receipts in a shard,
+  chunk-producers for that shard stop accepting more transactions.
+  https://github.com/near/nearcore/pull/9222
+
+But this does not put any limits on what shards will accept for other shards.
+For example, when a particular contract is popular, the contract's shard will
+eventually stop accepting new transactions but all other shards keep accepting
+more and more transactions that produce receipts for the popular contract.
+Therefore the number of delayed receipts keeps growing indefinitely.
+
+Cross-shard congestion control addresses this issue by stopping new transactions
+at the source and delaying receipt forwarding when the receiving shard has
+reached its congestion limit.
+
+## Specification
+
+The proposal adds fields to chunks headers, it adds a new trie column, it changes
+the rules to select transactions, and it changes the chunk execution flow. The
+next four sections specify each of these changes in more details.
+
+### Chunk header changes
+
+We change the chunk header to include congestion information.
+Specifically, we include two indicators.
+
+```rust
+ShardChunkHeaderInnerV3 {
+  // ...all fields from ShardChunkHeaderInnerV2
+
+  congestion_info: CongestionInfo
+}
+```
+
+Where alternatives for congestion info are decribed in more detail below in the 
+`Efficiently computing the congestion information` section. 
+
+Depending on the alternative solution this adds between 5 and 48 bytes to the
+chunk header, increasing it from 374 bytes up to 422 bytes in borsh representation.
+(Assuming no validator proposals included.) This in turn increases the block size
+by up to 48 bytes per shard.
+
+Including all this information in the chunk header enables efficient validation.
+Using either the previous chunk header or the state, and the list of receipts 
+applied and forwarded, a validator can check that the congestion rules described 
+in this NEP are fulfilled.
+
+### Changes to transaction selection
+
+Transaction selection is changed to reject new transactions when the system is
+congested, to reduce to total workload in the system.
+
+Today, transactions are taken from the chunk producer's pool until `tx_limit` is
+reached, where `tx_limit` is computed as follow.
+
+```python
+# old
+tx_limit = 500 Tgas if len(delayed_receipts) < 20_000 else 0
+```
+
+We replace the formula for the transaction limit to depend on the
+`incoming_congestion` computed in the previous shard variable (between 0 and 1):
```

Comment:
Do you mean computed in the previous chunk for this shard?

Reply:
Yes. The sentence should be fixed now to make more sense.